Build near filter when key is an array

viki53 opened the pull request:
Fix `buildNearFilter` to work with any key depth

### Description

When building a `$near` filter, keys can have different depth. To make the filter work properly, this needs to be handled properly.

So here's a fix for that.

NOTE: This comes after a fix for strongloop/loopback-datasource-juggler to handle deep `geo.nearFilter` search

#### Related issues

- Makes required strongloop/loopback-datasource-juggler#1216 (related to strongloop/loopback-datasource-juggler#993) fix work fully on MongoDB

### Checklist

- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

bajtos commented:
@viki53 thank you for the pull request.

In the description, you write:

> When building a $near filter, keys can have different depth. To make the filter work properly, this needs to be handled properly.

Please add a test making a query using keys with different depth, to verify your implementation and prevent regressions in the future.

viki53 commented:
@bajtos I provided a test but it seems to fail as the current data model does not use proper GeoJSON.  
But the query sent to the MongoDB server is the right one. Might need to create a new Model to test this case in depth at some point, as there are many possible edge cases…

I also edited the datasource juggler so that it doesn't provide an array as the key if it's a root key: this should help remain retro-compatible.

bajtos commented on package.json:
```diff
@@ -37,7 +37,7 @@
     "eslint": "^2.7.0",
     "eslint-config-loopback": "^1.0.0",
-    "loopback-datasource-juggler": "^3.0.0",
+    "loopback-datasource-juggler": "git+https://github.com/ebarault/loopback-datasource-juggler.git",
```
Please revert. If this pull request depends on changes from @ebarault's fork, then those changes must be landed first, before this pull request can be accepted.

bajtos commented on test/mongodb.test.js:
```diff
@@ -1567,7 +1568,7 @@ describe('mongodb connector', function() {
     });
 
-    it('find should be able to query by location', function(done) {
+    it.only('find should be able to query by location', function(done) {
```
forgotten `.only`, please remove

bajtos commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1612,67 @@ describe('mongodb connector', function() {
+          PostWithLocation.create({
+            title: 'Post #' + postNb,
+            location: new GeoPoint({
+              lat: coords.lat + (Math.random() * 2 - 1),
```
Using random data in tests make it difficult to reproduce and debug a test failure. Please use constant hard-coded values instead.

viki53 replied:
I removed the random value. Fixed coordinates shouldn't affect the behavior of the test so much.

bajtos commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1612,67 @@ describe('mongodb connector', function() {
+          createPostWithLocation,
+          createPostWithLocation,
+          createPostWithLocation,
+          createPostWithLocation,
```
Why are there exactly four posts created? It's a good practice to keep the test data as minimal as possible. Is it enough to have two model instances for this test?

viki53 replied:
It was useful in a previous version of the test, but it's not necessary any more. I'll remove that.

viki53 commented:
@bajtos All done!

bajtos commented:
> I provided a test but it seems to fail as the current data model does not use proper GeoJSON.
But the query sent to the MongoDB server is the right one. Might need to create a new Model to test this case in depth at some point, as there are many possible edge cases…

Yes please, try to cover at least the most important edge cases.

@jannyHou @superkhau you are more familiar with this codebase than I am, could one of you take over the code review please?

viki53 commented:
Just added a more robust test. I modified the `Superhero` model that wasn't of much use and re-used it for this use with a GeoJSON feature (with allows me to use a compound key — `location.geometry` — as well as depth with `$and`).

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
       });
     });
 
+    it.only('find should be able to query around location in deep/multiple keys', function(done) {
```
remove .only

superkhau commented on test/mongodb.test.js:
```diff
@@ -1531,6 +1540,7 @@ describe('mongodb connector', function() {
       PostWithLocation = geoDb.define('PostWithLocation', {
         _id: { type: geoDb.ObjectID, id: true },
         location: { type: GeoPoint, index: true },
+        title: { type: String },
```
This change seems unrelated, why the addition?

viki53 replied:
This was for the previous version of the test. Forgot about it. I'll delete it.

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
+          createSuperheroWithLocation,
+          createSuperheroWithLocation,
+        ], function(err) {
+          should.not.exist(err);
```
`if (err) return done(err);`

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
+              ],
+            },
+          }, function(err, results) {
+            should.not.exist(err);
```
if (err) return done(err);

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
+            },
+          }, function(err, results) {
+            should.not.exist(err);
+            should.exist(results);
```
no need for this check as you are verifying results in the next line

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
+          }, function(err, results) {
+            should.not.exist(err);
+            should.exist(results);
+            results.should.have.property('length', 1);
```
results.should.have.length(1);

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
+
+      geoDb.autoupdate(function(err) {
+        var heroNb = 0;
+        var powers = ['fly', 'lasers', 'strength', 'drink'];
```
Are these properties required for this test?

viki53 replied:
They help make sure every created Superhero is different, this way I make sure there is only one matching the query below.

superkhau replied:
Isn't the heroNumber enough though?

superkhau replied:
I see, you're using the index from line 82.

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
+        var powers = ['fly', 'lasers', 'strength', 'drink'];
+
+        var createSuperheroWithLocation = function(callback) {
+          heroNb++;
```
also remove this

superkhau commented on test/mongodb.test.js:
```diff
@@ -1611,6 +1621,74 @@ describe('mongodb connector', function() {
+
+          Superhero.create({
+            name: 'Hero #' + heroNb,
+            power: powers[heroNb - 1],
```
remove the two above properties as they add no value to the test

viki53 replied:
`power` is a required index. Plus it's used in the `$and` filter.

viki53 commented:
@superkhau Done!

Have a great weekend

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+  function searchForNear(node) {
+    if (!node) {
+      return;
+    }
```
add empty line here

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+      return;
+    }
+    if (Array.isArray(node)) {
+      node.forEach(function(nodeProperty, key) {
```
remove unused arg key

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+          found = true;
+        } else {
+          searchForNear(nodeProperty);
+        }
```
we should remove this duplication:

```

function nearFound(prop) {
  return prop && prop.near;
}
```

then in both cases above, you can just pass in:

```
Object.keys(node).forEach(function(key) {
  var prop = node[key];
  isFound = found(prop);
  if (!isFound)
    searchForNear(prop);
});
```

And reuse found() for the array check too

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+
+    if (typeof near.near === 'string') {
+      var s = near.near.split(',');
+      coordinates.lng = Number(s[0]);
```
coordinates.lng = parseInt(s[0], 10);
 coordinates.lng = parseInt(s[1], 10);

viki53 replied:
More like `parseFloat`, no? Digits matter *a lot* in coordinates.

superkhau replied:
Nice catch! ;) Yes, `parseFloat` please.

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+    }
+
+    var props = ['maxDistance', 'minDistance'];
+    //use mongodb default unit 'meters' rather than 'miles'
```
Nitpick, space in front of `use`

```
// Use mongodb...

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+
+    props.forEach(function(p) {
+      if (near[p]) {
+        queryValue['near']['$' + p] = convertToMeters(near[p], unit);
```
queryValue.near['$' + p] = ...

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+
+        property[near.key[i - 1]] = queryValue;
+      } else {
+        property = query.where[near.key] = queryValue;
```
why are we assigning to `= queryValue` at the end again?

viki53 replied:
The `property = ` part isn't very useful, it's just to stay coherent with the code above.

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,100 @@ function convertToMeters(distance, unit) {
+          if (near.key.hasOwnProperty(i + 1)) {
+            if (!property.hasOwnProperty(subKey)) {
+              property[subKey] = Number.isInteger(near.key[i + 1]) ? [] : {};
+            }
```
can we clean up these shortcuts a bit to make it more clear your intents? i'm having trouble understanding easily. something like:

```
var subKey = near.key[i];
var nextSubKeyPosition = i + 1;
var keyHasNextSubKey = near.key.hasOwnProperty(nextSubKeyPosition);
var propertyHasSubKey = property.hasOwnProperty(subKey);
if (hasNextSubKey && !propertyHasSubKey)
  property[subKey] = ...
```

etc, you're obviously manipulating the keys in some complicated way. I just wanna make sure when we see it again in 6 months that it'll be easy to decipher. ;)

viki53 replied:
I'd rather add somme comments to explain the code, if you'll allow it, as I'm not sure what you propose is mush easier to understand…

superkhau commented:
@bajtos The changes overall LGTM, do you want to take a final look?

bajtos commented on test/mongodb.test.js:
```diff
@@ -211,7 +220,7 @@ describe('mongodb connector', function() {
         var indexes =
         { _id_: [['_id', 1]],
-          geometry_2dsphere: [['geometry', '2dsphere']],
+          geojson_location_geometry: [['location.geometry', '2dsphere']],
```
IIUC, this change means our tests are no longer verifying `geometry_2dsphere`. I think we should preserve them, to ensure backwards compatibility is preserved. (This applies to the change above too.)

@superkhau thoughts?

viki53 replied:
I'm pretty sure it still does, I basically changed the name of the key. The key is still created with the right type.

bajtos replied:
oh, I see, makes sense.

bajtos commented on test/mongodb.test.js:
```diff
@@ -82,10 +82,19 @@ describe('mongodb connector', function() {
       power: { type: String, index: true, unique: true },
       address: { type: String, required: false, index: { mongodb: { unique: false, sparse: true }}},
       description: { type: String, required: false },
-      geometry: { type: Object, required: false, index: { mongodb: { kind: '2dsphere' }}},
```
oh, I see, makes sense.

bajtos replied:
@viki53 How about this one, can we preserve the index being defined in-line in the property definition?

viki53 replied:
Not sure, as the key is now a compound (`location.geometry`).

superkhau replied:
@viki53 Maybe we can avoid all the above as suggested by @bajtos by creating your own index (and model if necessary) instead of reusing Superhero? Thoughts?

viki53 replied:
I could, but that means creating even more data (which you wanted to avoid earlier) for little or no gain: the key is still created and properly checked so the test doesn't really change.

Even better: the test is called something like "creating complex indexes", so with a name like that it's even more complex and makes use of a different way of creating an index which isn't really tested elsewhere.

bajtos replied:
Yeah, let's keep things simple and use the version as it is now (as of ed4e581).

jannyHou commented:
sorry I was on vacation, will take a look asap.

superkhau commented:
@bajtos LGTM, LGTY?

@jannyHou Welcome back, do you want to work on landing this? I already reviewed/approved.

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,113 @@ function convertToMeters(distance, unit) {
+          if (near.key.hasOwnProperty(i + 1)) {
+            // if the query doesn't still have the current key set (undefined)
+            if (!property.hasOwnProperty(subKey)) {
+              // if the next level of depth is an integer, we create an array, otherwise an empty object
```
I was thinking more along the lines of a single comment explaining the block:

// if key is array, recursively search all items in array for subkeys

jannyHou commented:
@slnode test please

jannyHou commented:
@superkhau Thank you :) I triggered a CI test unfortunately node 4&6 are failing on the new test case:
http://ci.strongloop.com/job/loopback-connector-mongodb/node=6.x,os=windows/2026/console
@viki53 doesn't have access to CI report so I post the error here, it's related to the index, @viki53 can you take a look? Thanks. I think this test doesn't rely on the juggler pr.

Error log:
```
1) mongodb connector geo queries find should be queryable using locations with deep/multiple keys:
     MongoError: error processing query: ns=lb-ds-mongodb-test-2026.shTree: $and
    power == "strength"
    GEONEAR  field=location.geometry maxdist=1.79769e+308 isNearSphere=0
Sort: { _id: 1 }
Proj: {}
 planner returned error: unable to find index for $geoNear query
      at Function.MongoError.create (D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\error.js:31:11)
      at queryCallback (D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\cursor.js:212:36)
      at D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\connection\pool.js:455:18
      at _combinedTickCallback (internal/process/next_tick.js:67:7)
      at process._tickCallback (internal/process/next_tick.js:98:9)
  ---------------------------------------------
      at handleOperationCallback (D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\connection\pool.js:454:24)
      at D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\connection\pool.js:490:9
      at authenticateStragglers (D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\connection\pool.js:429:16)
      at Connection.messageHandler (D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\connection\pool.js:463:5)
      at Socket.<anonymous> (D:\workspace\loopback-connector-mongodb\d60c4a15\node_modules\mongodb-core\lib\connection\connection.js:319:22)
      at readableAddChunk (_stream_readable.js:176:18)
      at Socket.Readable.push (_stream_readable.js:134:10)
      at TCP.onread (net.js:548:20)
```

superkhau commented:
@jannyHou Thanks for taking care of this, please coordinate with @viki53 to figure out how to make it pass and ping me again then.

viki53 commented:
@jannyHou This error is indeed linked to the datasource juggler, as it is required to keep the key depth (the array).

I also found an other bug when running the tests again locally, I'm checking it out, I think I made a typo somewhere in my latest changes.

viki53 commented:
All done.

* Comment is updated as per @superkhau's recommandation
* Bug I found (distance calculation during results check ended up with NaN value) is fixed

Just needs to be tested with the right version of the datasource-juggler (see strongloop/loopback-datasource-juggler#1216)

superkhau commented on lib/mongodb.js:
```diff
@@ -768,45 +768,111 @@ function convertToMeters(distance, unit) {
+    var property;
+
+    if (near.key) {
+      // if key is an Array, set the $near query at the right depth, following the Array
```
@viki53 TY, much better

jannyHou commented:
@slnode test please

viki53 commented:
I didn't write the test for the juggler myself (they come from the original PR), so I'll need to look at it more closely.

viki53 commented:
@jannyHou Did you run those juggler tests with this version of the connector?

jannyHou commented:
@viki53 oh I see, thank you and I will review the juggler pr asap too. 

> Did you run those juggler tests with this version of the connector?

Yes.

ebarault commented:
@jannyHou : could you sync with @0candy so that https://github.com/strongloop/loopback-datasource-juggler/pull/1216 is landed together with the present one?

ssh24 commented:
@slnode test please

ebarault commented:
@ssh24 : i updated the PR to use the new mongoKey param returned by the juggler
i also rebased and squashed
all the tests pass locally

ssh24 commented:
@slnode test please

kjdelisle commented:
@slnode test please

ebarault commented:
CI failed for obscure reasons...

rmg commented:
@ebarault Travis needs updated build tools to support native node addons in v4+. #365 has been merged, so this PR can be rebased and those build failures should go away.

ssh24 commented:
@ebarault I already rebased for you.

ssh24 commented:
@ebarault The test case you added along with 10 others are failing on CI. Please take a look at it. https://travis-ci.org/strongloop/loopback-connector-mongodb/jobs/219397812

ebarault commented:
I guess that the CI build calls the juggler from npm? This PR requires the latest juggler from master.

kjdelisle commented:
@ebarault There's some CI infrastructure churn that might be causing some of these issues; normally master builds immediately get placed into a registry mirror for use in other PR builds, which doesn't seem to be happening. I'm looking into it now.

ssh24 commented:
@slnode test please

ssh24 commented:
Blocked on this till https://github.com/strongloop/loopback-datasource-juggler/pull/1314 is re-landed.

ssh24 commented:
PR is ready to land. @ebarault @viki53 Thank you for the contribution. Once this is merged, https://github.com/strongloop/loopback-datasource-juggler/pull/1314 will be ready to proceed with.

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

slnode commented:
Can one of the admins verify this patch?

ebarault commented:
Oh oh oh merged? Really? About time :-)
Thanks @ssh24 for finalizing

viki53 commented:
Thanks a lot for your patience, guys!
